### ISSUE TITLE
URI malformed error while fetching attrs

### ISSUE THREAD

**Issue by clayton13** [usnistgov/jsfive#33]
Hi!

I have been getting the following error when attempting to parse a specific file's attrs. Running latest jsfive-0.3.8 from relases.

It appears like something is off with my file (importing from another system outside of my control) as the library reports a string with length of 256 bytes where only "Second" is the value. I believe that the value field is null terminated and the library attempts to parse the full 256 bytes instead of up to the null, leading to a crash while trying to decode. 

Is this expected behavior for such cases? I would expect that if there is a null in a string that is not escaped, the string would terminate vs filtering the null and continuing. 

```
"URIError: URI malformed
    at decodeURIComponent (<anonymous>)
    at DataView64.getString (file:///C:/Users/claytonwhite/Documents/experiments/hdf5.js:207:14)
    at DataObjects._attr_value (file:///C:/Users/claytonwhite/Documents/experiments/hdf5.js:5407:34)
    at DataObjects.unpack_attribute (file:///C:/Users/claytonwhite/Documents/experiments/hdf5.js:5349:24)
    at DataObjects.get_attributes (file:///C:/Users/claytonwhite/Documents/experiments/hdf5.js:5279:34)
    at Group.get attrs [as attrs] (file:///C:/Users/claytonwhite/Documents/experiments/hdf5.js:6009:41)
    at FileReader.reader.onloadend (file:///C:/Users/claytonwhite/Documents/experiments/hd5viewer.html:204:42)"
```
```
"SecondàhîÖ&è.1>¥)Dáz¿ÿÿÿÿoû®	¯Ù»H*;òíÇU°±²ÇUT®ï3ÇUÖî8¯áé"
```

Relevant code here: 
https://github.com/usnistgov/jsfive/blob/64e0927148f3abf9347c2e3703e3850f73038b81/esm/core.js#L211-L221

**Comment by bmaranville** [comment]
There are potentially a couple things going on here... for one the use of ```decodeURIComponent``` here in the library is incorrect and it should be probably be using e.g. TextDecoder instead.  

I am also curious what the datatype of the offending field is.

If you can attach an HDF5 file that shows this problem, and indicate which path within the file points to the offending text field, I will try to fix it.

**Comment by clayton13** [comment]
Thanks for the supper speedy response!

There are a couple of seeming offending fields in the files I am working with, ideally I would like to access them both. I have shared them with you via email. 

Is [h5wasm](https://github.com/usnistgov/h5wasm) perhaps more fitting to use?

```js
// one with this URI malformed error
f.get("/Waveforms").values[0].attrs
// a "Uncaught not yet implemented!" for a table record?
f.get("/Frame").values[0].value
```

**Comment by bmaranville** [comment]
I will fix the handling of strings in jsfive because it has needed doing for a long time, anyway - but for your case the errors will still occur in both fields due to unimplemented features in jsfive.  

- The "Enumerated" datatype is not implemented in jsfive, and that is what can be found in e.g. f.get("/Waveforms/Channel 1").attrs["WaveformType"]
- The Compound datatype is not implemented in jsfive, and that is what is found in the "/Frame/TheFrame" dataset in your file

To answer your question about h5wasm, both of these features of HDF5 are (mostly) supported in the h5wasm library.  If you go to https://ncnr.nist.gov/ncnrdata/view/nexus-hdf-viewer.html you can click the "Load local file..." button at the bottom, and it indeed loads and parses the fields you are interested in, using h5wasm.  The Compound datatype is fully supported, and the Enumerated datatype seems to be half-working in the latest versions of h5wasm (it is returning an array instead of a single value), and I will look into that.

**Comment by clayton13** [comment]
Thank you for the explanation and background. That sounds good for others too. 

I really appreciate the h5wasm demo and these projects, all of which were really helpful in quickly getting me going.

I had some time to port over to h5wasm as it looked better suited for my use case. Along with supporting these fields of interest I am seeing a speed increase from ~10s to a speedy ~1.5s for ~100MB files in my data pipeline! 😃

**Comment by bmaranville** [comment]
I ended up adding support for the Enumerated type in jsfive, because it was just a few lines of code (it's just an integer when you read it), but the Compound type would be a lot more work.

I'm glad you're happy with h5wasm!  It will certainly be better supported going forward.  I'll close this issue since the string handling has been fixed in the latest release, v0.3.9 of jsfive.
